src/sdl2/common.rs

```diff
@@ -1,39 +1,102 @@
-use std::error::Error;
-use std::fmt;
+use std::ffi::{CString, NulError};
+use std::{error, fmt, io};
 
-/// A given integer was so big that its representation as a C integer would be
-/// negative.
-#[derive(Debug, Clone, PartialEq)]
-pub enum IntegerOrSdlError {
-    IntegerOverflows(&'static str, u32),
-    SdlError(String),
+#[derive(Debug)]
+pub struct SdlError {
+    msg: String,
 }
+
+impl SdlError {
+    pub fn from_last_error() -> Self {
+        let msg = crate::get_error();
+        Self { msg }
+    }
+
+    pub(crate) fn from_string(msg: String) -> Self {
+        Self { msg }
+    }
+
+    pub fn message(&self) -> &str {
+        &self.msg
+    }
+}
+
+impl fmt::Display for SdlError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.msg)
+    }
+}
+
+impl error::Error for SdlError {}
+
+#[derive(Debug)]
+pub enum Error {
+    Sdl(SdlError),
+    Io(io::Error),
+    /// An integer was larger than [`i32::MAX`] in a parameter, and it can't be converted to a C int
+    IntOverflow(&'static str, u32),
+    /// A null byte was found within a parameter, and it can't be sent to SDL
+    InvalidString(NulError, &'static str),
+}
+
+impl Error {
+    pub fn from_sdl_error() -> Self {
+        Self::Sdl(SdlError::from_last_error())
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Sdl(msg) => write!(f, "SDL error: {msg}"),
+            Self::Io(err) => write!(f, "IO error: {err}"),
+            Self::IntOverflow(name, value) => write!(f, "Integer '{name}' overflows: {value}"),
+            Self::InvalidString(name, nul) => write!(f, "Invalid string '{name}': {nul}"),
```

Comment:
In the definition the first field is the `NulError`, not the name.

To prevent these bugs (in this crate but also user code using this crate) I advise to use a struct like variant instead of a tuple like one. This also adds context to the context-less `&'static str` field. I suggest to change the definition to `InvalidString { nul_err: NulError, var_name: &'static str }` (or something like this) and roughly the same for `IntOverflow`.